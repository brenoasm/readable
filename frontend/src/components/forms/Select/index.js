import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import FormErrorMessage from '../FormErrorMessage';

const propTypes = {
  title: PropTypes.string,
  name: PropTypes.string,
  id: PropTypes.string,
  className: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  handleChange: PropTypes.func,
  errors: PropTypes.arrayOf(PropTypes.string),
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.value
  }))
};

const defaultProps = {
  title: '',
  name: '',
  id: '',
  className: '',
  value: '',
  handleChange: () => {},
  options: [],
  errors: []
}

const StyledSelect = styled.div`
  margin: 5px;

  & > label {
    margin-right: 5px;
  }

  & > select {
    min-height: 15px;
    min-width: 120px;
    padding: 1px 0;
    cursor: pointer;
  }
`;

const Select = ({
  title,
  name,
  id,
  className,
  value,
  handleChange,
  options,
  errors
}) => (
  <StyledSelect>
    <label htmlFor={name}>{title}</label>
    <select
      id={id}
      className={className}
      name={name}
      value={value}
      onChange={(event) => handleChange(event)}
    >
      <option key="default" value="">Selecione</option>
      {Array.isArray(options) &&
        options.map(option => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
    </select>
    <FormErrorMessage errors={errors} />
  </StyledSelect>
);

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
