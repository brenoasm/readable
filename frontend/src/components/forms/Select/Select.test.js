/* eslint-disable */

import React from 'react';

import Select from './';

describe('Select', () => {
  it('matches the snapshot', () => {
    const initialState = {
      name: 'category',
      title: 'categories',
      value: 'redux',
      id: 'category',
      handleChange: jest.fn(),
      errors: [],
      options: [
        {
          label: 'react',
          value: 'react'
        },
        {
          label: 'redux',
          value: 'redux'
        },
        {
          label: 'udacity',
          value: 'udacity'
        }
      ]
    };

    const wrapper = shallow(<Select {...initialState}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('shows a title when the property is passed', () => {
    const initialState = {
      title: 'categories'
    };

    const wrapper = shallow(<Select {...initialState}/>);

    expect(wrapper.find('label').text()).toBe(initialState.title);
  });

  it('should have a FormErrorMessage', () => {
    const initialState = {
      errors: [
        'Erro ao escolher opção'
      ],
    };

    const wrapper = shallow(<Select {...initialState}/>);

    expect(wrapper.find('FormErrorMessage')).toHaveLength(1);
  });

  it('trigger the handleChange event when the value changes', () => {
    const initialState = {
      handleChange: jest.fn(),
      options: [
        {
          label: 'react',
          value: 'react'
        },
        {
          label: 'redux',
          value: 'redux'
        },
        {
          label: 'udacity',
          value: 'udacity'
        }
      ]
    };

    const wrapper = shallow(<Select {...initialState}/>);

    wrapper.find('select').simulate('change', { target: { value: 'react' }});

    expect(initialState.handleChange).toHaveBeenCalled();
  });
});
