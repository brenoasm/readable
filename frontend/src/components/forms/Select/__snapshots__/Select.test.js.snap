// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    className=""
    errors={Array []}
    handleChange={[Function]}
    id="category"
    name="category"
    options={
        Array [
            Object {
              "label": "react",
              "value": "react",
            },
            Object {
              "label": "redux",
              "value": "redux",
            },
            Object {
              "label": "udacity",
              "value": "udacity",
            },
          ]
    }
    title="categories"
    value="redux"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <label
          htmlFor="category"
>
          categories
</label>,
        <select
          className=""
          id="category"
          name="category"
          onChange={[Function]}
          value="redux"
>
          <option
                    value=""
          >
                    Selecione
          </option>
          <option
                    value="react"
          >
                    react
          </option>
          <option
                    value="redux"
          >
                    redux
          </option>
          <option
                    value="udacity"
          >
                    udacity
          </option>
</select>,
        <FormErrorMessage
          errors={Array []}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "categories",
          "htmlFor": "category",
        },
        "ref": null,
        "rendered": "categories",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
>
              Selecione
</option>,
            Array [
              <option
                value="react"
>
                react
</option>,
              <option
                value="redux"
>
                redux
</option>,
              <option
                value="udacity"
>
                udacity
</option>,
            ],
          ],
          "className": "",
          "id": "category",
          "name": "category",
          "onChange": [Function],
          "value": "redux",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "default",
            "nodeType": "host",
            "props": Object {
              "children": "Selecione",
              "value": "",
            },
            "ref": null,
            "rendered": "Selecione",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "react",
            "nodeType": "host",
            "props": Object {
              "children": "react",
              "value": "react",
            },
            "ref": null,
            "rendered": "react",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "redux",
            "nodeType": "host",
            "props": Object {
              "children": "redux",
              "value": "redux",
            },
            "ref": null,
            "rendered": "redux",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "udacity",
            "nodeType": "host",
            "props": Object {
              "children": "udacity",
              "value": "udacity",
            },
            "ref": null,
            "rendered": "udacity",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errors": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <label
            htmlFor="category"
>
            categories
</label>,
          <select
            className=""
            id="category"
            name="category"
            onChange={[Function]}
            value="redux"
>
            <option
                        value=""
            >
                        Selecione
            </option>
            <option
                        value="react"
            >
                        react
            </option>
            <option
                        value="redux"
            >
                        redux
            </option>
            <option
                        value="udacity"
            >
                        udacity
            </option>
</select>,
          <FormErrorMessage
            errors={Array []}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "categories",
            "htmlFor": "category",
          },
          "ref": null,
          "rendered": "categories",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
>
                Selecione
</option>,
              Array [
                <option
                  value="react"
>
                  react
</option>,
                <option
                  value="redux"
>
                  redux
</option>,
                <option
                  value="udacity"
>
                  udacity
</option>,
              ],
            ],
            "className": "",
            "id": "category",
            "name": "category",
            "onChange": [Function],
            "value": "redux",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "default",
              "nodeType": "host",
              "props": Object {
                "children": "Selecione",
                "value": "",
              },
              "ref": null,
              "rendered": "Selecione",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "react",
              "nodeType": "host",
              "props": Object {
                "children": "react",
                "value": "react",
              },
              "ref": null,
              "rendered": "react",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "redux",
              "nodeType": "host",
              "props": Object {
                "children": "redux",
                "value": "redux",
              },
              "ref": null,
              "rendered": "redux",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "udacity",
              "nodeType": "host",
              "props": Object {
                "children": "udacity",
                "value": "udacity",
              },
              "ref": null,
              "rendered": "udacity",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errors": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
